lib/Doctrine/ORM/Internal/IdentityMap.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+namespace Doctrine\ORM\Internal;
+
+use Countable;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\ORMInvalidArgumentException;
+
+use function array_map;
+use function array_sum;
+use function get_class;
+use function implode;
+use function in_array;
+
+class IdentityMap implements Countable
+{
+    /** @var EntityManagerInterface */
+    private $entityManager;
+
+    public function __construct(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * Map of all identifiers of managed entities.
+     * Keys are object ids (spl_object_hash).
+     *
+     * @var mixed[]
+     * @psalm-var array<string, array<string, mixed>>
+     */
+    private $entityIdentifiers = [];
+
+    /**
+     * The identity map that holds references to all managed entities that have
+     * an identity. The entities are grouped by their class name.
+     * Since all classes in a hierarchy must share the same identifier set,
+     * we always take the root class name of the hierarchy.
+     *
+     * @var mixed[]
+     * @psalm-var array<class-string, array<string, object|null>>
+     */
+    private $identityMap = [];
+
+    public function addEntityIdentifier(string $oid, $identifier): void
```

Comment:
I found just one place where this method is used without adding entity to an identity map. And probably we can re-write the logic to use just identity map methods, without entity identifiers. Please check UnitOfWork::persistNew()

```php
            // Some identifiers may be foreign keys to new entities.
            // In this case, we don't have the value yet and should treat it as if we have a post-insert generator
            if (! $this->hasMissingIdsWhichAreForeignKeys($class, $idValue)) {
                $this->identityMap->addEntityIdentifier($oid, $idValue);
            }
```